### PR TITLE
feat: Add serving applications on subdomains and port-based proxying

### PR DESCRIPTION
**Pull request by Emyrk**
# What this does

This PR adds the ability to serve applications at `/` on subdomains. It also allows accessing ports inside a workspace with an anonymous application. 

Examples:

- Status Quo. An app on a path
  - https://dev.coder.com/@Emyrk/dev-dogfood/apps/code-server/
- **NEW** App on subdomain
  -  https://emyrk--dev-dogfood--main-code-server.dev.coder.com/
- **NEW** Port on path
  - https://dev.coder.com/@Emyrk/dev-dogfood/apps/8080/
- **NEW** Port on subdomain
  -  https://emyrk--dev-dogfood--main-8080.dev.coder.com/

# Conversation Wanted

I hit a wall working on this, and want some discussion before I proceed.

## Token Scoping + Security

We need to scope the `session_token` used for application authentication. It should only have access to application endpoints, and nothing else. This is so a malicious application cannot make authenticated requests on a user's behalf.

We can do this with RBAC token scopes as defined in the RFC, or by making a new token type like we did for agent authentication. The new token can live in `coder_application_token`. 

If the application is using the path based routing, then the `session_token` of the primary auth can still be used. We need to either have a security setting to always host applications on another subdomain, or somehow reject api calls from these paths (??).

## Applications on another domain

Eventually we want to serve applications on another domain for security reasons. To do this, we need to implement a way to insert a cookie on that subdomain, like we did with devurls in V1. I suggest we do this as a follow up PR.

## Applications on subpath

We should be checking if authenticated requests come from an app via the `Origin` header. If it is, we should not allow that Origin to make authenticated requests.

---------

# At present

At present we do not allow sharing these application URLs to other users. So all of the security considerations can be follow ups. 

----------

# Future Work

- Support serving apps on other domains
- Scoped tokens to just application authentication
- Security check on the things above, might need some setting toggles to disallow certain things.

### CODE REVIEW

**Comment by bpmct** [comment]
I'll defer to @sreya and @kylecarbs on the token scoping. From a product POV, supporting the following as a first iteration seems solid:

```sh
coder.mydomain.com # host
*.coder.mydomain.com # subdomains for apps
```

We can open another issue "Run Coder apps on a separate domain" as some users may eventually want

```sh
coder.mydomain.com
*.different-coder-apps-domain.com
```

(We have a few Coder Classic customers who do this today for security reasons)

**Comment by Emyrk** [comment]
> (We have a few Coder Classic customers who do this today for security reasons)

Yes, it's more secure to run from another domain. I think we will want to implement some sort of toggle on the BE to block the same domain in the future.

Awesome on first iteration though :+1:. For now, I think I'll just throw a cookie on both `.domain` and `domain` to handle this. We will need to tighten security in future.

File: coderd/subdomain.go
Comment (kylecarbs) [inline coderd/subdomain.go:32]: Seems like this should go in `httpmw` instead to follow existing patterns. Is there a reason it can't be that I'm missing?

**Review by kylecarbs — CHANGES_REQUESTED** [review]

File: coderd/subdomain.go
Comment (Emyrk) [reply coderd/subdomain.go:32]: Yea, the organization of this is a bit annoying.

The bottom calls `api.proxyWorkspaceApplication`. So the subdomain and path based code all use the app proxy code. If it's in `httpmw`, then I have to export that function and pass it into the `httpmw` function somehow?

That seemed a bit more annoying to read.

**Comment by bpmct** [comment]
For named applications, will this PR make the [relative_path](https://registry.terraform.io/providers/coder/coder/latest/docs/resources/app#relative_path) property for coder_app function as expected? (e.g. links from to apps from the web UI will go to the subdomain)

**Comment by Emyrk** [comment]
> For named applications, will this PR make the [relative_path](https://registry.terraform.io/providers/coder/coder/latest/docs/resources/app#relative_path) property for coder_app function as expected? (e.g. links from to apps from the web UI will go to the subdomain)

It will not. The deployment still needs to setup their certs with wildcards to support this feature. So `relative_path` still requires some deployment side config. We need to figure out how to go about displaying this :thinking:

File: coderd/subdomain.go
Comment (kylecarbs) [inline coderd/subdomain.go:87]: From my perspective, this parsing logic should entirely be placed in the `httpmw` package instead. It'll be a bit simpler to test, won't need to be exported since it'll be in the handler, and probably be small. Thoughts?

**Review by kylecarbs — DISMISSED** [review]

**Comment by kylecarbs** [comment]
@Emyrk since our `session_token` is HTTPOnly, is there any way to steal the cookie even with a malicious application? I'm thinking no, but I'm not sure.

File: coderd/subdomain.go
Comment (f0ssel) [reply coderd/subdomain.go:79]: I agree, going the other way is a little odd but I do think `8080--main--dev--f0ssel` is easier to read.

**Comment by f0ssel** [comment]
> @Emyrk since our `session_token` is HTTPOnly, is there any way to steal the cookie even with a malicious application? I'm thinking no, but I'm not sure.

We also strip the session token header on the proxy as well, correct?

File: coderd/subdomain.go
Comment (Emyrk) [reply coderd/subdomain.go]: I don't think I follow. Do you mean like an `ExtrackSubdomainAppURL` method to pull it out in the handler?

**Comment by Emyrk** [comment]
> @Emyrk since our `session_token` is HTTPOnly, is there any way to steal the cookie even with a malicious application? I'm thinking no, but I'm not sure.

We strip the cookie in the proxy app: https://github.com/coder/coder/blob/0578588bb7c4e9b8cb4bb627e097296735132bba/coderd/workspaceapps.go#L186-L188

There is something you can do though with a malicious app. You can send authenticated requests to coderd. This is an issue I put in the description. We should not allow sharing apps until we close that

File: coderd/subdomain.go
Comment (Emyrk) [reply coderd/subdomain.go]: Flipped :+1:

File: coderd/workspaceapps_test.go
Comment (kylecarbs) [reply coderd/workspaceapps_test.go:257]: Agreed

Comment (deansheather) [reply coderd/workspaceapps_test.go:260]: Kyle and I decided on `port/app--agent--workspace--user` to be in order of significance/hierarchy

**Review by BrunoQuaresma — APPROVED** [review]
LGTM

**Comment by BrunoQuaresma** [comment]
@deansheather right now, I'm doing this in the FE:

```
# Subdomain
`${port}--${agentName}--${workspaceName}--${username}.dev.coder.com`

# Subpath
`https://dev.coder.com/@${username}/${workspaceName}/apps/${port}`
```

Is that right?

**Comment by deansheather** [comment]
Yes @BrunoQuaresma, but the subpath should be `/@$username/$workspace.$agent/apps/$port`. The port is also interchangeable with an application name in both cases

File: coderd/workspaceapps.go
Comment (kylecarbs) [inline coderd/workspaceapps.go:291]: I _think_ `httpapi` might be a better spot for this too, but it's a bit weird, so chef's choice on this one. Regardless, I think it'd be better to remove this from being exposed and use an `_internal.go` test for it.

File: cli/server.go
Comment (kylecarbs) [inline cli/server.go:690]: I'm happy we stopped wainting

File: coderd/workspaceapps.go
Comment (kylecarbs) [inline coderd/workspaceapps.go:336]: This is interesting... I didn't know we could apply to subdomains without requiring an authentication redirect, this is awesome.

**Review by kylecarbs — COMMENTED** [review]
@deansheather are we going to use [`relative_path`}(https://registry.terraform.io/providers/coder/coder/latest/docs/resources/app#relative_path) to determine the client-side routing?

It seems like we'll need to add a server-side option for the wildcard path so we can redirect users on the frontend properly, or we could assume a wildcard above the root, but we'd just have to document that.

File: cli/server.go
Comment (kylecarbs) [reply cli/server.go:690]: shoutout @deansheather for being the biggest wainter I know

File: coderd/workspaceapps.go
Comment (deansheather) [reply coderd/workspaceapps.go:336]: We probably won't be doing this soon as we'll be implementing proper devurl auth similarly to v1

File: coderd/users.go
Comment (sreya) [inline coderd/users.go:894]: `api.setAuthCookie`

File: coderd/workspaceapps.go
Comment (deansheather) [reply coderd/workspaceapps.go:291]: how do you do the internal test thing? I've never seen that before

File: coderd/workspaceapps.go
Comment (sreya) [inline coderd/workspaceapps.go:170]: `http.Redirect` expects a URL but you've supplied a path here.

**Comment by deansheather** [comment]
> It seems like we'll need to add a server-side option for the wildcard path so we can redirect users on the frontend properly, or we could assume a wildcard above the root, but we'd just have to document that.

@kylecarbs this PR only does wildcards underneath the site `--access-url` (i.e. `dev.coder.com` => `app--agent--ws--user.dev.coder.com`), and not if the access URL is an IP address as that's obviously not supported.

In a future PR when we implement proper devurl auth (with their own tokens, similar to v1) we will be able to support separate domains, and it'll be configurable through a separate `--devurl-domain` flag or something.

This PR is the MVP for hostname based routing

**Review by sreya — APPROVED** [review]
Seems reasonable overall

**Review by bpmct — COMMENTED** [review]
The frontend uses an environment variable which decides whether links in the dashboard should go to a wildcard or path (if the admin didn't set up wildcard DNS, for example).

https://github.com/coder/coder/blob/38a0f7b4c383d6d4938c1e923666df450350a76c/site/webpack.common.ts#L17

What do you think about having this also impact the functionality (e.g. whether the app is routed) versus it being purely a cosmetic/dashboard variable?

If it's the latter, perhaps we should rename the var in #3812. I suggested the name, but originally assumed it'd be used across both.

File: coderd/workspaceapps.go
Comment (deansheather) [reply coderd/workspaceapps.go]: Done

File: coderd/workspaceapps.go
Comment (deansheather) [reply coderd/workspaceapps.go]: Moved to httpapi

File: coderd/workspaceapps.go
Comment (deansheather) [reply coderd/workspaceapps.go:167]: I don't think `http.Redirect` expects a full URL and works fine with relative or absolute paths. From godoc:
> Redirect replies to the request with a redirect to url, which may be a path relative to the request path.

Comment (sreya) [reply coderd/workspaceapps.go:167]: Ah that's nice, wasn't sure

Comment (deansheather) [reply coderd/workspaceapps.go:167]: I just tried this program to double check and it seems to work.

```go
package main

import (
	"fmt"
	"net/http"
)

func main() {
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		http.Redirect(w, r, "/hi", http.StatusTemporaryRedirect)
	})
	http.HandleFunc("/hi", func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprintf(w, "hi")
	})

	panic(http.ListenAndServe(":8080", nil))
}
```

**Comment by deansheather** [comment]
@bpmct the env var can be temporary until we do the changes I said in my above comment. I'll add a check for the env var

**Comment by deansheather** [comment]
Actually @bpmct that's a build time env var, not a proper one. I'm just going to merge this as is and we can figure out what we want to do about it when we solve the devurl auth problem

File: coderd/users_test.go
Comment (kylecarbs) [inline coderd/users_test.go:273]: Why do we need a custom `AccessURL` to check if multiple cookies are set? We can change `127.0.0.1` to localhost, which is a valid hostname, then we won't need to pass this option.

File: coderd/workspaceapps_test.go
Comment (kylecarbs) [inline coderd/workspaceapps_test.go:287]: Could we redirect to `/login` with a `redirect` query parameter back to the subdomain URL instead? Seems like this would be a questionable UX.

**Review by kylecarbs — COMMENTED** [review]
My comment mentioning `relative_path` indicates that in its current form, there's no way to access a wildcard application from the frontend. We should respect that value, and direct a user to the subdomain if it's specified (maybe it's called `wildcard` instead, so we don't break current behavior).

We also need to specify whether the wildcard is enabled or not, and have that information surfaced to the frontend so the "Open Port" button open in a subsequent PR functions.

**Comment by deansheather** [comment]
@kylecarbs Bruno is working on this in a separate frontend PR. He will be reading that value to figure out whether to link to wildcard or path-based route.

**Comment by BrunoQuaresma** [comment]
@kylecarbs here is the related PR https://github.com/coder/coder/pull/3812

File: coderd/workspaceapps_test.go
Comment (deansheather) [reply coderd/workspaceapps_test.go:287]: I've made it redirect to login but it's impossible to redirect back to the subdomain because we can't determine whether the accessed scheme was HTTP or HTTPS. If users are using a reverse proxy in front of coder (which they likely will) then coder will think it's receiving HTTP traffic when the actual scheme is HTTPS.

File: coderd/workspaceapps.go
Comment (Emyrk) [reply coderd/workspaceapps.go:243]: @kylecarbs this is a stop gap. It will not work on apps hosted on another domain. 

Also currently using the same full auth token. We should used scoped auth tokens.

**Review by kylecarbs — APPROVED** [review]

File: coderd/workspaceapps.go
Comment (deansheather) [reply coderd/workspaceapps.go:186]: We can't render the dashboard onto a devurls subdomain because we have a PWA app. It will try to make API requests to `app--agent--ws--user.domain.com/api/v2` which will fail. We also don't have a way of redirecting the user to the real access URL and showing an API error page, and we probably don't want to do that either because now if they refresh they'll be on the dashboard and not their devurl.

We'll need to add a small renderable error page similar to v1, but for now I don't think this is worth blocking this PR over. I made a ticket for it: #4040

Comment (kylecarbs) [reply coderd/workspaceapps.go:195]: I'm not sure what you mean by PWA app? I don't believe we have one right now.

Comment (deansheather) [reply coderd/workspaceapps.go:195]: i meant SPA oops

Comment (kylecarbs) [reply coderd/workspaceapps.go:195]: The SPA can detect if it's on a subdomain, right? It can just look at the host header and route accordingly.

Comment (deansheather) [reply coderd/workspaceapps.go:195]: We can't serve the dashboard app from subdomains like these because:

- It's comprised of multiple files (JS, CSS, images) and not just a single HTML file we can return (and these files are inaccessible on devurl subdomains)
- The dashboard assumes the API is at `/api/v2` when it's actually on a different domain as APIs are inaccessible from devurl subdomains
- The dasboard relies on a full session cookie for APIs to work, and we're planning on using a devurl-only session cookie

Comment (deansheather) [reply coderd/workspaceapps.go:195]: A single HTML file like we use in v1 means we don't need to worry about any of these things. All it needs to be is an error page with the coder logo, an error message, and two buttons

Comment (deansheather) [reply coderd/workspaceapps.go:195]: Spoke to Kyle on slack about this and we figured it out

**Pull request merged**
